src/crimson/os/seastore/journal/circular_bounded_journal.cc

```diff
@@ -0,0 +1,573 @@
+  if (encoded_size > get_available_size()) {
+    logger().error(
+      "CBJournal::submit_record: record size {}, but available size {}",
+      encoded_size,
+      get_available_size()
+      );
+    return crimson::ct_error::erange::make();
+  }
+
+  journal_seq_t j_seq {
+    header.cur_segment_seq++,
+    convert_abs_addr_to_paddr(
+      get_written_to(),
+      header.device_id)};
+  ceph::bufferlist to_write = encode_record(
+    std::move(record), device->get_block_size(),
+    j_seq, 0);
+  if (get_written_to() +
+      ceph::encoded_sizeof_bounded<record_group_header_t>() > header.end) {
+    // not enough space between written_to and the end of journal
+    set_used_size(get_used_size() + (header.end - get_written_to()));
+    set_written_to(get_start_addr());
+    if (encoded_size > get_available_size()) {
+      return crimson::ct_error::erange::make();
+    }
+  }
+  auto target = get_written_to();
+  if (get_written_to() + to_write.length() >= header.end) {
+    set_written_to(get_start_addr() +
+      (to_write.length() - (header.end - get_written_to())));
+  } else {
+    set_written_to(get_written_to() + to_write.length());
+  }
+  logger().debug(
+    "submit_record: mdlength {}, dlength {}, target {}",
+    r_size.get_mdlength(),
+    r_size.dlength,
+    target);
+
+  auto write_result = write_result_t{
+    j_seq,
+    (seastore_off_t)to_write.length()
+  };
+  auto write_fut = append_record(to_write, target);
+  return handle.enter(write_pipeline->device_submission
+  ).then([write_fut = std::move(write_fut)]() mutable {
+    return std::move(write_fut
+    ).handle_error(
+      write_ertr::pass_further{},
+      crimson::ct_error::assert_all{
+        "Invalid error in CBJournal::append_record"
+      }
+    );
+  }).safe_then([this, &handle] {
+    return handle.enter(write_pipeline->finalize);
+  }).safe_then([this, target,
+    length=to_write.length(),
+    &write_result,
+    &r_size] {
+    logger().debug(
+      "append_record: commit target {} used_size {} written length {}",
+      target, get_used_size(), length);
+
+    set_last_committed_record_base(target);
+    set_used_size(get_used_size() + length);
+    paddr_t paddr = convert_abs_addr_to_paddr(
+      target + r_size.get_mdlength(),
+      header.device_id);
+    auto submit_result = record_locator_t{
+      paddr,
+      write_result
+    };
+    return submit_result;
+  });
+}
+
+CBJournal::write_ertr::future<> CBJournal::device_write_bl(
+    rbm_abs_addr offset, bufferlist &bl)
+{
+  auto length = bl.length();
+  if (offset + length > header.end) {
+    return crimson::ct_error::erange::make();
+  }
+  logger().debug(
+    "overwrite in CBJournal, offset {}, length {}",
+    offset,
+    length);
+  auto write_length = length < get_block_size() ? get_block_size() : length;
+  auto bptr = bufferptr(ceph::buffer::create_page_aligned(write_length));
+  auto iter = bl.cbegin();
+  iter.copy(bl.length(), bptr.c_str());
```

Comment:
See bufferlist::rebuild_aligned

Reply:
added rebuild_aligned above.

Reply:
Yes, which means that these lines are not necessary.  bufferlist::rebuild_aligned updates bl to contain only aligned bufferptrs -- you can write them directly without allocating/copying a new one.